[#9] 회원 탈퇴기능 구현

msugo1 opened the pull request:
## Controller 계층 관련

### UserController 

- 사용자의 회원 탈퇴 요청을 받을 userDeleteAccount 메소드와endpoint 추가

## Service 계층 관련

### UserService

- 사용자의 회원 탈퇴 요청에 대한 비즈니스 로직을 처리할
  userDeleteAccount 메소드 추가

1. 비밀번호가 일치하지 않으면 `InvalidUserInfoException` 예외 발생

2. 회원 탈퇴처리 완료 후 로그아웃 처리

## Repository 계층 관련

### UserRepository/UserMapper

- 탈퇴 요청 처리 전 비밀번호 재확인을 위해 사용자 아이디를 가지고 DB에
  등록된 비밀번호를 리턴하는 `getUserPasswordById` 메소드 및 해당 Query 추가

- 해당 유저에 대한 논리 삭제를 진행할 `deleteUser` 메소드 추가 및 해당
  Query 추가 (isDeactivate를 Y로 변경)

- `getUserById` Query 처리 시 `isDeactivate`가 `Y가 아닌` 회원들만
  DB에서 가져오도록 변경

## 테스트 관련

- 위에서 작성한 로직들을 검증하기 위한 테스트 작성 및 실행 (완료)

- `Postman`을 활용해 해당 기능들이 제대로 작동하는지 재확인

Commit:
[#9] 회원 탈퇴기능 구현

- 사용자의 회원 탈퇴 요청을 받을 userDeleteAccount 메소드와endpoint 추가

- 사용자의 회원 탈퇴 요청에 대한 비즈니스 로직을 처리할
  userDeleteAccount 메소드 추가

1. 비밀번호가 일치하지 않으면 `InvalidUserInfoException` 예외 발생

2. 회원 탈퇴처리 완료 후 로그아웃 처리

- 탈퇴 요청 처리 전 비밀번호 재확인을 위해 사용자 아이디를 가지고 DB에
  등록된 비밀번호를 리턴하는 `getUserPasswordById` 메소드 및 해당 Query 추가

- 해당 유저에 대한 논리 삭제를 진행할 `deleteUser` 메소드 추가 및 해당
  Query 추가 (isDeactivate를 Y로 변경)

- `getUserById` Query 처리 시 isDeactivate가 Y가 아닌 회원들만
  DB에서 가져오도록 변경

- 위에서 작성한 로직들을 검증하기 위한 테스트 작성 및 실행 (완료)

- `Postman`을 활용해 해당 기능들이 제대로 작동하는지 재확인

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -60,4 +61,12 @@ public void userFindPassword(@Valid @RequestBody UserFindPasswordRequest findPas
         userService.findUserPassword(findPasswordRequest);
     }
+
+    @DeleteMapping("/account")
```
`DELETE /users/account` 라고 하면 어떤 리소스에 대해 어떤 동작을 하는지 명확하지 않습니다.
탈퇴라고 하면 `내 정보를 삭제하는 것` 이라고도 볼 수 있을 것 같은데 어떻게 표현해야 좋을까요?

msugo1 replied:
결국 계정을 삭제하는 것도 `내 정보`를 취급하는 것과 관련이 있는 듯 하여

`MyPageController`로 이동하여 `URL`을 `DELETE /my-infos/account` 로 변경해주었습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserService.java:
```diff
@@ -89,4 +88,26 @@ public void findUserPassword(UserFindPasswordRequest findPasswordRequest) {
 
+    /*
+        계정 삭제 요청을 받으면 해당 사용자의 정보를 바로 삭제하는 것이 아니라 일정기간 동안 보관했다가 삭제한다고 하여 물리삭제가 아니라
+        계정 비활성화 정보를 'Y'로 바꾸는 식으로 하여 논리삭제를 구현하였습니다.
```
👍

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserService.java:
```diff
@@ -89,4 +88,26 @@ public void findUserPassword(UserFindPasswordRequest findPasswordRequest) {
+        계정 비활성화 정보를 'Y'로 바꾸는 식으로 하여 논리삭제를 구현하였습니다.
+
+        아이디 삭제 처리 후 로그아웃을 할 때 LoginService 을 주입받아 logout 을 사용하면 순환참조가 발생하여 HttpSession 을 주입받아
+        session invalidation 으로 로그아웃 처리를 대신하였습니다.
```
이거는 좀 애매하네요~ `유저 데이터를 관리`를 맡은 곳에서 세션의 존재를 알면 안돼는것 아닐까요?

f-lab-dev commented on src/test/java/me/soo/helloworld/repository/UserRepositoryTest.java:
```diff
@@ -129,4 +129,39 @@ public void userUpdateProfileImageTestSuccess() {
+    }
+
+    @Test
+    @DisplayName("현재 사용자의 정보를 DB에서 삭제합니다.")
```
테스트를 작성할때는 실패하는 케이스를 더 신경써서 작성해야합니다~
`어떨 때는 탈퇴하는게 실패한다` 이런 경우도 추가해주면 좋을 것 같아요~

msugo1 replied:
`Service Layer` 테스트에는 두 가지 예외관련 테스트 실패 케이스
 - 계정 비밀번호와 요청받은 비밀번호가 다른 경우
 - 혹시나 중복 삭제요청이 들어오는 경우 
 
추가했고,

`Repository Layer` 테스트에는 성공 케이스 이외에도

- 이미 삭제 처리된 회원에 대해 재삭제 요청이 들어오는 경우 

에 대한 실패 케이스 추가하였습니다.

Commit:
[#9] 피드백 반영

Controller 계층

* UserController (delete)

- userAccountDelete 메소드 삭제

: 계정 삭제는 결국 `내 정보 삭제`와 관련이 크다고 생각하여
MyPageController로 옮기기 위해 UserController 내에서 메소드 삭제

* MyPageController (new)

- myAccountDelete 메소드 추가

- URL 변경 (기존: `DELETE /users/account` -> `DELETE /my-infos/account)

테스트 추가

* 실패 케이스에 대한 테스트 추가

- Service Layer Test

1. 계정 비밀번호와 요청받은 `비밀번호가 다른 경우`

2. 혹시나 `중복 삭제요청`이 들어오는 경우

- Repository Layer Test

1. 이미 삭제 처리된 회원에 대해 재삭제 요청이 들어오는 경우

* 이전 branch로 부터의 변경사항을 활용하기 위한 merge
- Merge branch 'feature/8' into feature/9

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/MyPageController.java:
```diff
@@ -20,6 +21,8 @@
 
     private final UserService userService;
 
+    private final HttpSession httpSession;
```
이런 세션관련 기능을 인증으로 캡슐화시키는건 `LoginService`로 해야하지 않을까요?

Commit:
[#9] 피드백 반영

* MyPageController
- 회원탈퇴 로직 완료 후 httpSession을 직접적으로 사용해서 세션을 invalidate 시키는 대신 loginService를 주입받아 로그아웃 기능을 사용하도록 변경
